### PR TITLE
Add container static types

### LINKED ISSUE #870: Implement `StaticType()` for arrays and dictionaries
### Issue To Be Solved

`getType()` function on containers (arrays. dictionaries), returns no type information.
e.g:
```kotlin
var array : [Int] = [1, 3]

log(array.getType())
```

Actual output: `Type()`

Expected Output: `Type([Int])`

### Solution

- [x] Checker: Record static types for array expressions and dictionary expressions in elaboration 
      @turbolent 
      #1034
- [x] Interpreter: 
    @turbolent 
    #1034
  - [x] Add static type field to array values and dictionary values
  - [x] Implement `StaticType()` for arrays and dictionaries
  - [x] Add static type information when evaluation array expressions and dictionary expressions (use elaboration)
- [x] Add function to walk value
  @turbolent 
  #1037
- [ ] Encoding and decoding
  - This requires a storage format change
  - [x] Add new encoder/decoder version
    @SupunS
    #1039
  - [x] Store static types 
    @SupunS
    #1035
  - [x] Read static types
    @SupunS 
    #1036
  - [x] Ensure existing stored data can be decoded using the old decoder and re-encoding using the new encoder: Arrays and dictionaries are first deferred and need to be forced to be decoded, and the new decoding functions for deferred arrays and dictionaries need to delegate to the old deferred decoding functions. + Tests
    @SupunS
    #1042
    #1046
  - [ ] Storage migration. Infer static types from stored values, by walking decoded value object graph (see item "Add function to walk value" above)
    @turbolent 
    - Similar to https://github.com/onflow/flow-go/blob/master/cmd/util/ledger/migrations/storage_v4.go
    - [x] Handle deferred values
    - [x] Migrate values even when type information is unavailable
    - [ ] Delete values that don't have type information
    - [ ] Generate report for deleted data
- [x] Import and export
  - [x] Add array type to imported array value and dictionary value
    @turbolent 
    #1034
  - [x] Add static type to imported array values and dictionary values using target parameter type
     @turbolent 
     #1038
  - [x] Infer type from the value when parameter/target type is not specific enough (e.g: target type has `AnyStruct` and the value is `[3, 5, 2]`). Could make use of: #1025, by getting static type of each element -> convert to sema type -> find super type -> convert back to static type.
   @SupunS
   Short-term solution: #1052
   Long-term solution: #1049
- [x] Dynamic types: Add static types to array values and dictionary values
  @turbolent 
  #1043
- [x] Dynamic subtype check: Consider the array's / dictionary's static type.
  @turbolent
  #1043
- [x] Update `ConformsToDynamicType`: Consider dynamic array type's / dynamic dictionary type's static type
  @SupunS
  Seems redundant. Added tests: #1045

### Concerns

- Q: How do we maintain backward compatibility as discussed in #195?
  A: We will infer static types for existing stored data and migrate it

### Related Issues

https://github.com/onflow/cadence/issues/748

### PR DESCRIPTION
**Pull request by SupunS**
Closes #870

## Description

Add static type information to containers (arrays and dictionaries)

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/value.go [hunk 1]
```diff
@@ -680,6 +683,8 @@ func NewDeferredArrayValue(
 	version uint16,
 ) *ArrayValue {
 	return &ArrayValue{
+		// TODO: type
```
Comment (turbolent) [inline runtime/interpreter/value.go:686]: Looks like we missed a TODO here. Could you please resolve it?

File: runtime/interpreter/value.go [hunk 2]
```diff
@@ -7354,6 +7432,8 @@ func NewDeferredDictionaryValue(
 	version uint16,
 ) *DictionaryValue {
 	return &DictionaryValue{
+		// TODO: type
```
Comment (turbolent) [inline runtime/interpreter/value.go:7435]: Looks like we missed a TODO here for dictionaries too. Could you please resolve it?

**Review by turbolent — COMMENTED** [review]
Looks like there are two, similar, TODOs that I think we should resolve before merging.

Other than that, I think we already reviewed all the PRs individually, so don't need a full re-review of it all, and a quick look over it also looks good.

Can't wait to ⛵ this!

File: runtime/interpreter/value.go (hunk 1 above)
Comment (SupunS) [reply runtime/interpreter/value.go:686]: Ah right! We can simply remove the TODO (+ field initializing). Static type is populated when the content is loaded to memory, so we dont need to provide it at the constructor. Same for dictionary as well.

Comment (turbolent) [reply runtime/interpreter/value.go:686]: Ah, right! 👍

Comment (SupunS) [reply runtime/interpreter/value.go]: Removed it with a comment: 5b2c6ce

Comment (turbolent) [reply runtime/interpreter/value.go]: Thank you for resolving them and also adding comments! 👌

**Review by turbolent — APPROVED** [review]
🎉

**Pull request merged**